### PR TITLE
ENH: Add keepdims argument to count_nonzero

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
A simple enhancement to count_nonzero:
```
In [3]: a
Out[3]: 
array([[ 0,  1,  7,  0],
       [ 3,  0,  2, 19]])

In [4]: np.count_nonzero(a, axis=1, keepdims=True)
Out[4]: 
array([[2],
       [3]])
```

### CODE REVIEW

**Comment by WarrenWeckesser** [comment]
The TravisCI test failure on the s390x platform looks unrelated to this PR.

**Comment by WarrenWeckesser** [comment]
Reran the tests and all passed--looks like the TravisCI test failure on the s390x platform was a transient.

**Comment by eric-wieser** [comment]
Needs a release note, but this looks like an obviously sane addition.

**Comment by WarrenWeckesser** [comment]
Thanks @eric-wieser, I added a release note.

File: doc/release/upcoming_changes/15870.new_feature.rst [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+------------------------------------------------
+The parameter ``keepdims`` was added to `numpy.count_nonzero`. The
+parameter has the same meaning as it does in reduction functions such
+as `numpy.sum` or `numpy.mean`.  When an ``axis`` argument is given,
+the slice along that dimension (or dimensions, if ``axis`` is a tuple
+with length greater than one) is reduced to a single number in the
+output.  Normally this means the shape of the output array has fewer
+dimensions than the input.  When ``keepdims=True`` is used, the output
+array has the same number of dimensions as the input, and the lengths
+of the reduced dimensions are one.
```
Comment (eric-wieser) [inline doc/release/upcoming_changes/15870.new_feature.rst]: Not sure if this section is needed, the user can look at the docs if they want the meaning.

Will let the person merging decide

**Review by eric-wieser — APPROVED** [review]

File: doc/release/upcoming_changes/15870.new_feature.rst (hunk 1 above)
Comment (WarrenWeckesser) [reply doc/release/upcoming_changes/15870.new_feature.rst]: Thanks @eric-wieser. I don't mind paring down the release note, but I'll wait for more feedback before copy-editing.

Comment (WarrenWeckesser) [reply doc/release/upcoming_changes/15870.new_feature.rst]: On second thought, I agree with the suggestion, so I just cut the release note down to the first two sentences.

File: numpy/core/numeric.py [hunk 2]
```diff
@@ -384,12 +384,12 @@ def full_like(a, fill_value, dtype=None, order='K', subok=True, shape=None):
     return res
 
 
-def _count_nonzero_dispatcher(a, axis=None):
+def _count_nonzero_dispatcher(a, axis=None, *, keepdims=False):
```
Comment (seberg) [inline numpy/core/numeric.py]: @shoyer shouldn't this be always `None` for the dispatching signature, did it actually matter?

File: numpy/core/numeric.py [hunk 3]
```diff
@@ -429,13 +436,17 @@ def count_nonzero(a, axis=None):
...
+    >>> np.count_nonzero(a, axis=0)
+    array([1, 1, 2, 1])
+    >>> np.count_nonzero(a, axis=1)
     array([2, 3])
-
+    >>> np.count_nonzero(a, axis=1, keepdims=True)
+    array([[2],
+           [3]])
     """
     if axis is None:
```
Comment (seberg) [inline numpy/core/numeric.py]: This seems incorrect, it has to be `axis is None and not keepdims`. Also needs a test.

**Review by seberg — COMMENTED** [review]
LGTM, just the `None` path needs fixing. Nice generalization, I agree that this is a reduce-like operation and thus should have the `keepdims` argument. kwarg-only is a nice touch.

File: numpy/core/numeric.py (hunk 2 above)
Comment (eric-wieser) [reply numpy/core/numeric.py]: I think it was just convention to avoid defaults getting out of sync. Still worthwhile though.

Comment (WarrenWeckesser) [reply numpy/core/numeric.py]: OK, I changed the `keepdims` default in the dispatcher function signature to None.

File: numpy/core/numeric.py (hunk 3 above)
Comment (WarrenWeckesser) [reply numpy/core/numeric.py]: Good catch, fixed.

File: numpy/core/numeric.py (hunk 2 above)
Comment (shoyer) [reply numpy/core/numeric.py]: Hmm, it looks like our import-time check to verify default values of dispatcher arguments may be broken on keyword only argument.

**Comment by seberg** [comment]
Thanks, lets put this in!

**Pull request merged**